### PR TITLE
planner, ddl: fix hash partition with adding new column

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: ref #39570

Problem Summary:
colsInfo used FullHiddenColsAndVisibleCols and can get the writeOnly/writeReorg column, result in `information of column `new column` is not found`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -288,7 +288,8 @@ func (s *partitionProcessor) pruneHashPartition(ctx sessionctx.Context, tbl tabl
 // please see https://github.com/pingcap/tidb/issues/22635 for more details.
 func (s *partitionProcessor) reconstructTableColNames(ds *DataSource) ([]*types.FieldName, error) {
 	names := make([]*types.FieldName, 0, len(ds.TblCols))
-	colsInfo := ds.table.FullHiddenColsAndVisibleCols()
+	// Use DeletableCols to get all the columns.
+	colsInfo := ds.table.DeletableCols()
```
Comment (bb7133) [inline planner/core/rule_partition_processor.go:292]: How about the other partition types?

Comment (wjhuang2016) [reply planner/core/rule_partition_processor.go:292]: It's used only by hash partition.

File: ddl/db_test.go [hunk 2]
```diff
@@ -1763,3 +1763,23 @@ func TestDDLBlockedCreateView(t *testing.T) {
...
+	tk.MustExec("create table t(a int, b int) partition by hash(a) partitions 4")
+
+	hook := &ddl.TestDDLCallback{Do: dom}
+	hook.OnJobRunBeforeExported = func(job *model.Job) {
+		if job.SchemaState != model.StateWriteOnly {
+			return
+		}
+		tk2 := testkit.NewTestKit(t, store)
+		tk2.MustExec("use test")
+		tk2.MustExec("delete from t")
```
Comment (tangenta) [inline ddl/db_test.go:1781]: Please use a variable to receive the error and assert it out of the hook. Otherwise, the test may timeout.

**Review by hawkingrei — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**